File: perl5/sync_bulk.pl
#!/usr/bin/perl
# sync_bulk.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/sync_bulk.py
# auch in: OpenClaw@gateway2:skills/sync-utils/scripts/sync_bulk.py
# Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use File::Find;
use File::stat;
use Time::Local;
use Getopt::Long;

# Bulk Sync - Synchronisiert alle Skills

# Add path for sync functions
unshift @INC, '/home/openclaw/.openclaw/workspace/scripts';

require "sync_clawhub_git.pl";

my $CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
my $GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";

sub sync_all_skills {
    my ($dry_run) = @_;
    
    # Alle Skills finden
    my %all_skills;
    
    if (-d $CLAWHUB_DIR) {
        opendir(my $dh, $CLAWHUB_DIR) or die "Cannot open directory $CLAWHUB_DIR: $!";
        while (readdir $dh) {
            next if /^\./;
            my $path = "$CLAWHUB_DIR/$_";
            if (-d $path) {
                $all_skills{$_} = 1;
            }
        }
        closedir $dh;
    }
    
    if (-d $GIT_DIR) {
        opendir(my $dh, $GIT_DIR) or die "Cannot open directory $GIT_DIR: $!";
        while (readdir $dh) {
            next if /^\./;
            my $path = "$GIT_DIR/$_";
            if (-d $path) {
                $all_skills{$_} = 1;
            }
        }
        closedir $dh;
    }
    
    my $count = keys %all_skills;
    &log("Bulk Sync: $count Skills gefunden");
    
    my %results = (
        "synced" => [],
        "skipped" => [],
        "failed" => []
    );
    
    for my $skill (sort keys %all_skills) {
        my $clawhub_path = "$CLAWHUB_DIR/$skill";
        my $git_path = "$GIT_DIR/$skill";
        
        eval {
            # Nur in ClawHub → zu Git
            if (-d $clawhub_path && !-d $git_path) {
                if (&validate_skill($clawhub_path)) {
                    &log("Syncing $skill to Git...");
                    if (&sync_to_git($skill, $dry_run)) {
                        push @{$results{"synced"}}, "$skill → Git";
                    } else {
                        push @{$results{"failed"}}, $skill;
                    }
                } else {
                    push @{$results{"skipped"}}, "$skill (validation failed)";
                }
            }
            
            # Nur in Git → zu ClawHub
            elsif (-d $git_path && !-d $clawhub_path) {
                if (&validate_skill($git_path)) {
                    &log("Syncing $skill to ClawHub...");
                    if (&sync_to_clawhub($skill, $dry_run)) {
                        push @{$results{"synced"}}, "$skill → ClawHub";
                    } else {
                        push @{$results{"failed"}}, $skill;
                    }
                } else {
                    push @{$results{"skipped"}}, "$skill (validation failed)";
                }
            }
            
            # In beiden - prüfe ob Update nötig
            elsif (-d $clawhub_path && -d $git_path) {
                # Vereinfachte Prüfung
                my $clawhub_mtime = get_max_mtime($clawhub_path);
                my $git_mtime = get_max_mtime($git_path, sub { $_[0] !~ /\.git/ });
                
                if (abs($clawhub_mtime - $git_mtime) > 60) {
                    if ($clawhub_mtime > $git_mtime) {
                        &log("Updating $skill in Git...");
                        if (&sync_to_git($skill, $dry_run)) {
                            push @{$results{"synced"}}, "$skill → Git (update)";
                        } else {
                            push @{$results{"failed"}}, $skill;
                        }
                    } else {
                        &log("Updating $skill in ClawHub...");
                        if (&sync_to_clawhub($skill, $dry_run)) {
                            push @{$results{"synced"}}, "$skill → ClawHub (update)";
                        } else {
                            push @{$results{"failed"}}, $skill;
                        }
                    }
                } else {
                    push @{$results{"skipped"}}, "$skill (already synced)";
                }
            }
        };
        if ($@) {
            &log("Error processing $skill: $@", "ERROR");
            push @{$results{"failed"}}, $skill;
        }
    }
    
    # Zusammenfassung
    print "\n" . "=" x 60 . "\n";
    print "Bulk Sync " . ($dry_run ? "DRY-RUN" : "EXECUTED") . " - Zusammenfassung\n";
    print "=" x 60 . "\n";
    print "✅ Synchronisiert: " . scalar(@{$results{"synced"}}) . "\n";
    for my $item (@{$results{"synced"}}) {
        print "   - $item\n";
    }
    print "\n⏭️  Übersprungen: " . scalar(@{$results{"skipped"}}) . "\n";
    if (scalar(@{$results{"skipped"}}) <= 10) {
        for my $item (@{$results{"skipped"}}) {
            print "   - $item\n";
        }
    } else {
        print "   - " . scalar(@{$results{"skipped"}}) . " Skills (bereits synchron oder Validierung fehlgeschlagen)\n";
    }
    print "\n❌ Fehlgeschlagen: " . scalar(@{$results{"failed"}}) . "\n";
    for my $item (@{$results{"failed"}}) {
        print "   - $item\n";
    }
    print "=" x 60 . "\n";
}

sub get_max_mtime {
    my ($dir, $filter) = @_;
    $filter //= sub { 1 };
    
    my $max_time = 0;
    find(sub {
        return unless -f $_;
        return unless $filter->($_);
        my $mtime = (stat($_))[9];
        $max_time = $mtime if $mtime > $max_time;
    }, $dir);
    
    return $max_time;
}

sub main {
    my ($dry_run, $execute);
    GetOptions(
        "dry-run" => \$dry_run,
        "execute" => \$execute
    ) or die "Falsche Optionen\n";
    
    if (!$dry_run && !$execute) {
        print "Bitte --dry-run oder --execute angeben\n";
        exit 1;
    }
    
    sync_all_skills($dry_run);
}

main() if not caller;

File: javascript/sync_bulk.js
#!/usr/bin/env node
// sync_bulk.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/sync_bulk.py
// auch in: OpenClaw@gateway2:skills/sync-utils/scripts/sync_bulk.py
// Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

/**
 * Bulk Sync - Synchronisiert alle Skills
 */

const fs = require('fs');
const path = require('path');

// Import sync functions
const {
  syncToGit,
  syncToClawhub,
  log,
  validateSkill
} = require('/home/openclaw/.openclaw/workspace/scripts/sync_clawhub_git.js');

const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";

function getMaxMtime(dirPath, excludePattern = null) {
  let maxMtime = 0;
  
  function walkDir(currentPath) {
    const entries = fs.readdirSync(currentPath, { withFileTypes: true });
    
    for (const entry of entries) {
      const fullPath = path.join(currentPath, entry.name);
      
      if (excludePattern && excludePattern.test(fullPath)) {
        continue;
      }
      
      if (entry.isDirectory()) {
        walkDir(fullPath);
      } else if (entry.isFile()) {
        const stat = fs.statSync(fullPath);
        if (stat.mtimeMs > maxMtime) {
          maxMtime = stat.mtimeMs;
        }
      }
    }
  }
  
  walkDir(dirPath);
  return maxMtime;
}

function syncAllSkills(dryRun = true) {
  // Alle Skills finden
  const allSkills = new Set();
  
  if (fs.existsSync(CLAWHUB_DIR)) {
    const clawhubDirs = fs.readdirSync(CLAWHUB_DIR, { withFileTypes: true })
      .filter(d => d.isDirectory() && !d.name.startsWith('.'))
      .map(d => d.name);
    clawhubDirs.forEach(d => allSkills.add(d));
  }
  
  if (fs.existsSync(GIT_DIR)) {
    const gitDirs = fs.readdirSync(GIT_DIR, { withFileTypes: true })
      .filter(d => d.isDirectory() && !d.name.startsWith('.'))
      .map(d => d.name);
    gitDirs.forEach(d => allSkills.add(d));
  }
  
  log(`Bulk Sync: ${allSkills.size} Skills gefunden`);
  
  const results = {
    "synced": [],
    "skipped": [],
    "failed": []
  };
  
  [...allSkills].sort().forEach(skill => {
    const clawhubPath = path.join(CLAWHUB_DIR, skill);
    const gitPath = path.join(GIT_DIR, skill);
    
    try {
      // Nur in ClawHub → zu Git
      if (fs.existsSync(clawhubPath) && !fs.existsSync(gitPath)) {
        if (validateSkill(clawhubPath)) {
          log(`Syncing ${skill} to Git...`);
          if (syncToGit(skill, dryRun)) {
            results.synced.push(`${skill} → Git`);
          } else {
            results.failed.push(skill);
          }
        } else {
          results.skipped.push(`${skill} (validation failed)`);
        }
      }
      
      // Nur in Git → zu ClawHub
      else if (fs.existsSync(gitPath) && !fs.existsSync(clawhubPath)) {
        if (validateSkill(gitPath)) {
          log(`Syncing ${skill} to ClawHub...`);
          if (syncToClawhub(skill, dryRun)) {
            results.synced.push(`${skill} → ClawHub`);
          } else {
            results.failed.push(skill);
          }
        } else {
          results.skipped.push(`${skill} (validation failed)`);
        }
      }
      
      // In beiden - prüfe ob Update nötig
      else if (fs.existsSync(clawhubPath) && fs.existsSync(gitPath)) {
        // Vereinfachte Prüfung
        const clawhubMtime = getMaxMtime(clawhubPath);
        const gitMtime = getMaxMtime(gitPath, /\.git/);
        
        if (Math.abs(clawhubMtime - gitMtime) > 60000) {
          if (clawhubMtime > gitMtime) {
            log(`Updating ${skill} in Git...`);
            if (syncToGit(skill, dryRun)) {
              results.synced.push(`${skill} → Git (update)`);
            } else {
              results.failed.push(skill);
            }
          } else {
            log(`Updating ${skill} in ClawHub...`);
            if (syncToClawhub(skill, dryRun)) {
              results.synced.push(`${skill} → ClawHub (update)`);
            } else {
              results.failed.push(skill);
            }
          }
        } else {
          results.skipped.push(`${skill} (already synced)`);
        }
      }
    } catch (e) {
      log(`Error processing ${skill}: ${e}`, "ERROR");
      results.failed.push(skill);
    }
  });
  
  // Zusammenfassung
  console.log("\n" + "=".repeat(60));
  console.log(`Bulk Sync ${dryRun ? 'DRY-RUN' : 'EXECUTED'} - Zusammenfassung`);
  console.log("=".repeat(60));
  console.log(`✅ Synchronisiert: ${results.synced.length}`);
  results.synced.forEach(item => console.log(`   - ${item}`));
  console.log(`\n⏭️  Übersprungen: ${results.skipped.length}`);
  if (results.skipped.length <= 10) {
    results.skipped.forEach(item => console.log(`   - ${item}`));
  } else {
    console.log(`   - ${results.skipped.length} Skills (bereits synchron oder Validierung fehlgeschlagen)`);
  }
  console.log(`\n❌ Fehlgeschlagen: ${results.failed.length}`);
  results.failed.forEach(item => console.log(`   - ${item}`));
  console.log("=".repeat(60));
}

function main() {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run');
  const execute = args.includes('--execute');
  
  if (!dryRun && !execute) {
    console.log("Bitte --dry-run oder --execute angeben");
    process.exit(1);
  }
  
  syncAllSkills(dryRun);
}

main();
